README.org
#+title: Readme

* StarFormatter
Star Formatter is a tool to convert json or csv data into starintel's json format.
It takes a json config file and uses that to know what fields contain the data types, like name, email, ect.
Arrays are supported, nested json is not at this point.

*** Compile & Install
Install nim lang.
~nimble install~ from project root.
*** Usage
Options
#+begin_example
Usage:
  main [REQUIRED,optional-params]
Options:
  -h, --help                                  print this cligen-erated help
  --help-syntax                               advanced: prepend,plurals,..
  -c=, --config=    string  "config.json"     set config
  -m=, --mode=      string  "json"            set mode
  --href=           string  "http:127.0.0.1"  set href
  -d=, --database=  string  "star-intel"      set database
  --couchPort=      int     5489              set couchPort
  -u, --upload      bool    false             set upload
  -o=, --output=    string  ""                set output
  -i=, --input=     string  REQUIRED          set input


#+end_example

Parse json

#+begin_src shell
starf -m json -i input.json -o ouput.json
#+end_src

Parse a csv and upload to a remote host
#+begin_src shell
starf -c cvs-config.json -m csv --href http://exampleCouchdb -i input.csv
#+end_src
*** Configuration
Star Formatter is configured by a json file that defines what fields contains what.
Examples will be put in the examples/ dir
Example config
#+begin_src json
{
    "jsonConfig": {
        "meta": {
            "id": "id",
            "rev": "rev",
            "dateAdded": "date",
            "dateUpdated": "date",
            "dataset": "database",
            "sourceDataset": "Google board of directors",
            "memberships": "",
            "operationId": "",
            "defaultOrgType": "Corperation"
        },
        "people": {
            "dtype": "person",
            "name": "",
            "fname": "fname",
            "mname": "mname",
            "lname": "lname",
            "email": "email",
            "phone": "",
            "orgName": "org",
            "orgType": "",
            "phoneArray": "",
            "emailArray": "",
            "orgArray": "",
            "street": "",
            "street2": "",
            "city": "",
            "postal": "",
            "region": "",
            "country": "",
            "address": "",
            "nameOrder": [],
            "roles": "roles",
            "title": "title",
            "bio": "",
            "socialMediaArray": "",
            "lat": "",
            "long": "",
            "ip": "",
            "dob": "",
            "gender": "",
            "interests": ""
        },
        "orgs": {
            "dtype": "org",
            "name": "name",
            "orgType": "type",
            "reg": "reg"
        },
        "emails": {
            "dtype": "email",
            "emailUsername": "emailUsername",
            "emailDomain": "emailDomain",
            "emailPassword": "emailPassword",
            "username": "username"
        },
        "address": {
            "dtype": "type",
            "street": "street",
            "street2": "street2",
            "city": "city",
            "region": "region",
            "zip": "zip",
            "country": "country"
        }
    }
}

#+end_src
**** Config options
***** Meta Options
| Option         | Description                                                                      |
|                |                                                                                  |
|----------------+----------------------------------------------------------------------------------|
| id             | This field will be used for document id, not required.                           |
| rev            | Used only if you are importing documents you exported from couchdb, not required |
| dateAdded      | Used to date when the document was added to the database                         |
| dateUpdated    | Used to date when the document was updated                                       |
| dataset        | What is the destination dataset?                                                 |
| sourceDataset  | Source for the dataset                                                           |
| operationId    | Operation id/ name                                                               |
| memberships    | Deprecated/not added. Might be used to define relationships                      |
| defaultOrgType | The Default the organization type (eg: NGO, Non Profit)                          |
|                |                                                                                  |

***** People Options

| Option           | Description                                                                |
|------------------+----------------------------------------------------------------------------|
| dtype            | used incase it is not a custom data type. Should probaly just use "person" |
| name             | Not used yet, but will be for the full name field                          |
| fname            | First name field                                                           |
| mname            | Middle name if needed                                                      |
| lname            | last name field                                                            |
| email            | email field.                                                               |
| emailArray       | email field, must be an array.                                             |
| phone            | phone field                                                                |
| phoneArray       | phone field, must be an array                                              |
| orgArray         | org array field, must be an array                                          |
| orgName          | Field with org name.                                                       |
| orgType          | Type of org, if not found will use the type in defaultOrgType              |
| address          | set to true if address data is needed                                      |
| street           | street address field                                                       |
| street2          | field with extra street info                                               |
| city             | city field                                                                 |
| region           | region/state field                                                         |
| country          | country field                                                              |
| postal           | postal code/zip field options                                              |
| roles            | role of the person at the organization                                     |
| nameOrder        | not used                                                                   |
| title            | not used                                                                   |
| bio              | think of it as an extra infomation field.                                  |
| socialMediaArray | an array of social media                                                   |
| lat              | lattitude, used for address                                                |
| long             | longtiude, used for address                                                |
| ip               | Needs to be renamed, its an array of ip field.                             |
| dob              | date of birth field                                                        |
| gender           | male/female field                                                          |
| interests        | an array of interests                                                      |

#+end_lang                                                         |


changelog.org
* Changelog
** [2023-04-07 Fri 02:10] 0.3.0
Updated formater to bring it inline with the removal of nested objects (starintel_doc 0.5.2)


src/starFormater.nim
import starintel_doc
import jsony, json
import parsecsv
import starintel_couchdb
import asyncdispatch
import mycouch
proc injestCsv*(config: MetaConfig, star: AsyncStarintelDatabase, path: string) {.async.} =
  var
    p: CsvParser
    docs: seq[JsonNode]
    jdoc: JsonNode
  p.open(path)
  p.readHeaderRow
  while p.readRow():
    try:
      var person = config.parsePerson(p)
      person.makeUUID
      jdoc = %*person

      jdoc["_id"] = jdoc["id"]
      docs.add(jdoc)
      jdoc = %*{}
      if docs.len == 1000:
        discard await star.server.bulkDocs(star.database, %*docs)
        docs = @[]
      if docs.len != 0:
        discard await star.server.bulkDocs(star.database, %*docs)
    except KeyError:
      discard
    except CsvError:
      discard
proc injestJson*(config: MetaConfig, star: AsyncStarintelDatabase, path: string) {.async.} =
  var
    docs: seq[JsonNode]
    jdoc: JsonNode
  let f = open(path, fmRead)
  for line in f.lines:
    try:
      var person = config.parsePerson(line.fromJson)
      jdoc = %*person
      jdoc["_id"] = jdoc["id"]
      docs.add(jdoc)
      jdoc = %*{}
      if docs.len == 1000:
        discard await star.server.bulkDocs(star.database, %*docs)
        docs = @[]
      if docs.len != 0:
        discard await star.server.bulkDocs(star.database, %*docs)
    except KeyError:
      discard
proc convertJson*(config: MetaConfig, input: string, output = "") =
  ## Read a File and output to a spec complient document
  let f = open(input, fmRead)
  var o: File
  o = open(output, fmAppend)
  for line in f.lines:
    try:
      var person = config.parsePerson(line.fromJson)
      var jdoc = %*person
      jdoc["_id"] = jdoc["id"]

      o.writeLine(jdoc)
      jdoc = %*{}
    except KeyError:
      discard
proc convertCsv*(config: MetaConfig, input: string, output="") =
  var
    p: CsvParser
    docs: seq[JsonNode]
    jdoc: JsonNode
    o: File
  p.open(input)
  o = open(output, fmAppend)
  p.readHeaderRow
  while p.readRow():
    try:
      var person = config.parsePerson(p)
      var jdoc = %*person
      jdoc["_id"] = jdoc["id"]

      o.writeLine(jdoc)
      jdoc = %*{}
    except KeyError:
      discard
    except CsvError:
      discard

proc main(config = "config.json", mode="json", href="http:127.0.0.1", database="star-intel", couchPort=5489, upload=false, output = "", input: string) =
  echo 1
  var star = initStarIntel(href, database, couchPort)
  let meta = readConfig(config)
  case mode:
    of "json":
      if upload == true:
        waitFor meta.injestJson(star, input)
      else:
        meta.convertJson(input, output)
    of "csv":
      if upload == true:
        waitFor meta.injestCsv(star, input)
      else:
        meta.convertCsv(input, output)

import cligen; dispatch main


src/starf/parsing.nim
import starintel_doc
import json
import parsecsv
import times
import strutils
type
  People* = object
    ## Configuration for json data holding data about people
    ## Fields are used to hold the name of the field
    fname*: string
    lname*: string
    mname*: string
    # if you only have a name field
    name*: string
    email*: string
    region*: string
    phone*: string # TODO
    phoneArray*: string # TODO
    bio*: string
    socialMediaArray*: string # TODO
    socialMedia*: string # TODO
    dob*: string # TODO
    gender*: string
    race*: string

  Orgs* = object
    ## Configuration for json data holding organization data
    name*: string # field holding name of org
    orgType*: string # field holding type of org NGO, corp, Government, ect
    reg*: string # registration number field
    bio*: string # About/bio page field
    website*: string
    country*: string
    defaultOrgType*: string # default org type
  Emails* =  object
    email*: string
    ## If you have the email username, domain and password
    emailUsername*: string
    emailDomain*: string
    emailPassword*: string
  Address* =  object
    ## Configuration for json data holding address data
    street*: string
    city*: string
    state*: string
    postal*: string
    street2*: string
    country*: string
    lat*: string
    long*: string
    alt*: string
  Metadata* =  object
    ## Not to Be confused with MetaConfig, this is the object holding metadata like dates and dataset info
    dateAdded*: string
    dateUpdated*: string
    dataset*: string
  MetaConfig* =  object
    ## Meta config holding all configrations for import jobs
    people*: People
    org*: Orgs
    address*: Address
    email*: Emails
    metadata*: Metadata

proc readConfig*(path: string): MetaConfig =
  ## Read a Json config that defines the mapping between fields
  var meta = MetaConfig()
  let f = open(path, fmRead)
  defer: f.close
  let jconfig = f.readAll.parseJson
  meta.metadata = jconfig["meta"].to(Metadata)
  meta.people = jconfig["people"].to(People)
  meta.org = jconfig["orgs"].to(Orgs)
  meta.address = jconfig["address"].to(Address)
  result = meta


proc nowUnix*(): int64 =
  result = now().toTime().toUnix()

proc parsePerson*(config: MetaConfig, line: JsonNode): BookerPerson =
  ## Parses a json line taking a MetaConfig object maping the fields from the people field.
  ##
  let defaultTime = nowUnix()
  var person = BookerPerson(dtype: "person", dataset: config.metadata.dataset)
  person.fname = line{config.people.fname}.getStr("")
  person.lname = line{config.people.lname}.getStr("")
  person.mname = line{config.people.mname}.getStr("")
  person.bio = line{config.people.bio}.getStr("")
  person.region = line{config.people.region}.getStr("")
  person.dob = line{config.people.dob}.getStr("")
  person.gender = line{config.people.gender}.getStr("")
  person.race = line{config.people.race}.getStr("")
  person.date_added = line{config.metadata.dateAdded}.getBiggestInt(defaultTime)
  person.date_updated = line{config.metadata.dateUpdated}.getBiggestInt(defaultTime)
  result = person

proc parsePerson*(config: MetaConfig, csv: var CsvParser): BookerPerson =
  let defaultTime = nowUnix()
  var person = BookerPerson(dtype: "person", dataset: config.metadata.dataset)
  person.fname = csv.rowEntry(config.people.fname)
  person.mname = csv.rowEntry(config.people.mname)
  person.lname = csv.rowEntry(config.people.lname)
  if config.people.region.len == 0 :
    person.region = csv.rowEntry(config.people.region)
  if config.metadata.dateAdded.len != 0:
    person.date_added = csv.rowEntry(config.metadata.dateAdded).parseBiggestInt()
  else:
    person.date_added = defaultTime

  if config.metadata.dateUpdated.len != 0:
    person.date_updated = csv.rowEntry(config.metadata.dateUpdated).parseBiggestInt()
  else:
    person.date_updated = defaultTime
  result = person


proc parseOrg*(config: MetaConfig, line: JsonNode): BookerOrg =
  let defaultTime = nowUnix()
  var org = BookerOrg(dtype: "org", dataset: config.metadata.dataset)
  org.name = line{config.org.name}.getStr("")
  org.reg = line{config.org.reg}.getStr("")
  org.bio = line{config.org.bio}.getStr("")
  org.website = line{config.org.website}.getStr("")
  org.country = line{config.org.country}.getStr("")
  org.date_added = line{config.metadata.dateAdded}.getBiggestInt(defaultTime)
  org.date_updated = line{config.metadata.dateUpdated}.getBiggestInt(defaultTime)
  if config.org.orgType.len != 0:
    org.etype = line{config.org.orgType}.getStr("")
  else:
    org.etype = config.org.defaultOrgType
  result = org

proc parseOrg*(config: MetaConfig, csv: var CsvParser): BookerOrg =
  let defaultTime = nowUnix()
  var org = BookerOrg(dtype: "org", dataset: config.metadata.dataset)
  org.name = csv.rowEntry(config.org.name)
  if config.org.reg.len != 0:
    org.reg = csv.rowEntry(config.org.reg)
  if config.org.bio.len != 0:
    org.bio = csv.rowEntry(config.org.bio)
  if config.org.website.len != 0:
    org.website = csv.rowEntry(config.org.website)
  if config.org.country.len != 0:
    org.country = csv.rowEntry(config.org.country)


  if config.metadata.dateAdded.len != 0:
    org.date_added = csv.rowEntry(config.metadata.dateAdded).parseBiggestInt()
  else:
    org.date_added = defaultTime

  if config.metadata.dateUpdated.len != 0:
    org.date_updated = csv.rowEntry(config.metadata.dateUpdated).parseBiggestInt()
  else:
    org.date_updated = defaultTime

  if config.org.orgType.len != 0:
    org.etype = csv.rowEntry(config.org.orgType)
  else:
    org.etype = config.org.defaultOrgType
  result = org


proc parseEmail*(config: MetaConfig, line: JsonNode): BookerEmail =
  let defaultTime = nowUnix()
  var email = BookerEmail(dtype: "email", dataset: config.metadata.dataset)
  if config.email.email.len != 0:
    email = newEmail(line{config.email.email}.getStr(), line{config.email.emailPassword}.getStr(""))
  else:
    email.emailUsername = line{config.email.emailUsername}.getStr()
    email.emailDomain = line{config.email.emailDomain}.getStr()
    email.emailPassword = line{config.email.emailPassword}.getStr("")
  email.date_added = line{config.metadata.dateAdded}.getBiggestInt(defaultTime)
  email.date_updated = line{config.metadata.dateUpdated}.getBiggestInt(defaultTime)
  result = email

proc parseEmail*(config: MetaConfig, csv: var CsvParser): BookerEmail =
  let defaultTime = nowUnix()
  var email = BookerEmail(dtype: "email", dataset: config.metadata.dataset)
  if config.email.email.len != 0:
    let e = csv.rowEntry(config.email.email)
    email = e.newEmail()
    if config.email.emailPassword.len != 0:
       email.emailPassword = csv.rowEntry(config.email.emailPassword)
  else:
    email.emailUsername = csv.rowEntry(config.email.emailUsername)
    email.emailDomain = csv.rowEntry(config.email.emailDomain)
    if config.email.emailPassword.len != 0:
       email.emailPassword = csv.rowEntry(config.email.emailPassword)
  if config.metadata.dateAdded.len != 0:
    email.date_added = csv.rowEntry(config.metadata.dateAdded).parseBiggestInt()
  else:
    email.date_added = defaultTime

  if config.metadata.dateUpdated.len != 0:
    email.date_updated = csv.rowEntry(config.metadata.dateUpdated).parseBiggestInt()
  else:
    email.date_updated = defaultTime
  result = email



proc parseAddress*(config: MetaConfig, line: JsonNode): BookerAddress =
  var address: BookerAddress
  address.street = line{config.address.street}.getStr("")
  address.street2 = line{config.address.street2}.getStr("")
  address.city = line{config.address.city}.getStr("")
  address.postal = line{config.address.postal}.getStr("")
  address.state = line{config.address.state}.getStr("")
  address.country = line{config.address.country}.getStr("")
  address.lat = line{config.address.lat}.getFloat(0.0)
  address.long = line{config.address.long}.getFloat(0.0)
  address.alt = line{config.address.alt}.getFloat(0.0)
  result = address


proc parseAddress*(config: MetaConfig, csv: var CsvParser): BookerAddress =
  var address: BookerAddress
  address.street = csv.rowEntry(config.address.street)
  address.street2 = csv.rowEntry(config.address.street2)
  address.city = csv.rowEntry(config.address.city)
  address.postal = csv.rowEntry(config.address.postal)
  address.state = csv.rowEntry(config.address.state)
  address.country = csv.rowEntry(config.address.country)
  if config.address.lat.len != 0 or config.address.long.len != 0:
    address.lat = csv.rowEntry(config.address.lat).parseFloat()
    address.long = csv.rowEntry(config.address.long).parseFloat()
    address.alt = csv.rowEntry(config.address.alt).parseFloat()
  result = address


src/starf.nim
import starintel_doc
import json
import parsecsv
import asyncdispatch
import mycouch
import starfpkg/parsing

proc parseCsvRow(config: MetaConfig, csv: var CsvParser): seq[JsonNode] =
  var
    docs: seq[JsonNode]
  if config.person.enabled:
    docs.add(config.parsePerson(csv).dump)
  if config.address.enabled:
    docs.add(config.parseAddress(csv).dump)
  if config.org.enabled:
    docs.add(config.parseOrg(csv).dump)
  if config.email.enabled:
    docs.add(config.parseEmail(csv).dump)
  result = config.link(docs)


proc parseJsonLine(config: MetaConfig, line: JsonNode): seq[JsonNode] =
  var docs: seq[JsonNode]
  if config.person.enabled:
    docs.add(config.parsePerson(line).dump)
  if config.address.enabled:
    docs.add(config.parseAddress(line).dump)
  if config.org.enabled:
    docs.add(config.parseOrg(line).dump)
  if config.email.enabled:
    docs.add(config.parseEmail(line).dump)
  result = config.link(docs)

proc genConfig(): JsonNode =
  let person = People(fname: "first_name", mname: "middle_name",
                      lname: "last_name", race: "race", gender: "male",
                      dob: "dob", bio: "bio", region: "region", enabled: true)
  let org = Orgs(name: "org_name", orgType: "org-type", reg: "reg", website: "website",
                 country: "United States", defaultOrgType: "corp", enabled: true)
  let emails = Emails(email: "email", emailUsername: "email_username", emailDomain: "email_domain", emailPassword: "password", enabled: false)
  let address = Address(street: "street", street2: "street2", country: "country", state: "state", postal: "postal",
                        city: "city", lat: "lat", long: "long", alt: "alt", enabled: true)
  let mapping = LinkMap(dtype: "person", targetType: "org", relation: Relations.member)
  var meta = MetaData(dataset: "star-intel", dateAdded: "date", dateUpdated: "date")
  meta.mapping.add(mapping)
  result = %*MetaConfig(person: person, org: org, email: emails, metadata: meta, address: address)
proc injestCsv*(config: MetaConfig, star: AsyncCouchDBClient, path: string, database: string = "star-intel") {.async.} =
  var
    p: CsvParser
    docs: seq[JsonNode]
  p.open(path)
  p.readHeaderRow
  while p.readRow():
    try:
      docs.add(config.parseCsvRow(p))
      if docs.len == 1000:
        discard await star.bulkDocs(database, %*docs)
        docs = @[]
      if docs.len != 0:
        discard await star.bulkDocs(database, %*docs)
    except KeyError:
      discard
    except CsvError:
      discard


proc injestJson*(config: MetaConfig, star: AsyncCouchDBClient, path: string, database: string = "star-intel") {.async.} =
  var docs: seq[JsonNode]
  let f = open(path, fmRead)
  for line in f.lines:
    try:
      docs.add(config.parseJsonLine(line.parseJson))
      if docs.len == 1000:
        discard await star.bulkDocs(database, %docs)
        docs = @[]
      if docs.len != 0:
        discard await star.bulkDocs(database, %docs)
    except KeyError:
      discard
proc convertJson*(config: MetaConfig, input: string, output = "") =
  ## Read a File and output to a spec complient document
  let f = open(input, fmRead)
  var o: File
  o = open(output, fmAppend)
  for line in f.lines:
    try:
      let docs = config.parseJsonLine(line.parseJson)
      for doc in docs:
        o.writeLine(doc)
    except KeyError:
      discard
proc convertCsv*(config: MetaConfig, input: string, output="") =
  var
    p: CsvParser
    o: File
  p.open(input)
  o = open(output, fmAppend)
  p.readHeaderRow
  while p.readRow():
    try:
      let docs = config.parseCsvRow(p)
      for doc in docs:
        o.writeLine(doc)
    except KeyError:
      discard
    except CsvError:
      discard


proc main(config = "config.json", mode="json", href="http:127.0.0.1",
                                      database="star-intel",
                                      couchPort=5489, upload=false,
                                      username = "", pass = "", output = "", input = "") =
  var star = newAsyncCouchDBClient(href, couchPort)
  case mode:
    of "json":
      let meta = readConfig(config)
      if upload == true:
        discard waitFor star.cookieAuth(username, pass)
        waitFor meta.injestJson(star, input, database)
      else:
        meta.convertJson(input, output)
    of "csv":
      let meta = readConfig(config)
      if upload == true:
        discard waitFor star.cookieAuth(username, pass)
        waitFor meta.injestCsv(star, input, database)
      else:
        meta.convertCsv(input, output)
    of "gen-config":
      let defaultConfig = %*MetaConfig()
      let o = open(config, fmWrite)
      defer: o.close()
      o.write(genConfig().pretty())
    else:
      echo "Unknown mode"
import cligen; dispatch main
